File: backend/app/schemas/labour.py
import uuid
from datetime import date, datetime
from typing import List, Optional
from pydantic import BaseModel, ConfigDict, Field


class AttendanceCreate(BaseModel):
    worker_id: str = Field(..., max_length=64, description="Worker identifier or badge number")
    worker_name: str = Field(..., max_length=255, description="Worker full name")
    mine_site_id: uuid.UUID
    contractor_id: Optional[uuid.UUID] = None
    shift_date: datetime
    shift_type: str = Field(default="day", description="Shift type: day, night, morning, evening")
    clock_in: datetime
    clock_out: Optional[datetime] = None
    hours_worked: Optional[float] = None
    overtime_hours: Optional[float] = None


class AttendanceOut(BaseModel):
    id: uuid.UUID
    worker_id: str
    worker_name: str
    mine_site_id: uuid.UUID
    contractor_id: Optional[uuid.UUID] = None
    shift_date: datetime
    shift_type: str
    clock_in: datetime
    clock_out: Optional[datetime] = None
    hours_worked: float
    overtime_hours: float
    is_violation: bool
    violation_reason: Optional[str] = None
    created_at: datetime

    model_config = ConfigDict(from_attributes=True)


class LabourRuleOut(BaseModel):
    id: uuid.UUID
    mine_site_id: Optional[uuid.UUID] = None
    max_shift_hours: float
    max_overtime_hours: float
    min_rest_hours_between_shifts: float
    updated_at: datetime

    model_config = ConfigDict(from_attributes=True)


class LabourRuleUpdate(BaseModel):
    max_shift_hours: Optional[float] = Field(None, ge=4.0, le=16.0)
    max_overtime_hours: Optional[float] = Field(None, ge=0.0, le=8.0)
    min_rest_hours_between_shifts: Optional[float] = Field(None, ge=8.0, le=24.0)


class LabourViolationsSummary(BaseModel):
    total_shifts: int
    active_workers: int
    violation_shifts: int
    overtime_breaches: int
    rest_breaches: int
    compliance_rate_pct: float

File: dashboard/src/api/labour.ts
import apiClient from './client';

export interface WorkerItem {
  id: string;
  badge_number: string;
  name: string;
  role: string;
  contractor_id?: string | null;
  contractor_name?: string | null;
  mine_site_id: string;
  mine_site_name?: string | null;
  is_active: boolean;
  created_at: string;
}

export interface WorkerCreatePayload {
  badge_number: string;
  name: string;
  role: string;
  contractor_id?: string | null;
  mine_site_id: string;
  is_active?: boolean;
}

export interface WorkerUpdatePayload {
  name?: string;
  role?: string;
  contractor_id?: string | null;
  is_active?: boolean;
}

export interface AttendanceItem {
  id: string;
  worker_id: string;
  worker_badge_number: string;
  worker_name: string;
  worker_role?: string | null;
  contractor_id?: string | null;
  contractor_name?: string | null;
  mine_site_id: string;
  shift_date: string;
  shift_type: string;
  clock_in: string;
  clock_out?: string | null;
  hours_worked: number;
  overtime_hours: number;
  is_violation: boolean;
  violation_reason?: string | null;
  created_at: string;
}

export interface AttendanceCreatePayload {
  worker_id: string;
  shift_date: string;
  shift_type: string;
  clock_in: string;
  clock_out?: string | null;
  hours_worked?: number | null;
  overtime_hours?: number | null;
}

export interface AttendanceBulkPayload {
  worker_ids: string[];
  mine_site_id: string;
  shift_date: string;
  shift_type: string;
  clock_in: string;
  clock_out?: string | null;
}

export interface LabourViolationsSummary {
  total_shifts: number;
  active_workers: number;
  violation_shifts: number;
  overtime_breaches: number;
  rest_breaches: number;
  compliance_rate_pct: number;
}

export interface LabourRule {
  id: string;
  mine_site_id?: string | null;
  max_shift_hours: number;
  max_overtime_hours: number;
  min_rest_hours_between_shifts: number;
  updated_at: string;
}

// ---------------------------------------------------------------------------
// Worker Master Directory Endpoints
// ---------------------------------------------------------------------------

export const fetchWorkers = async (params?: {
  mine_site_id?: string;
  query?: string;
  is_active?: boolean;
  limit?: number;
}): Promise<WorkerItem[]> => {
  const res = await apiClient.get<WorkerItem[]>('/workers', { params });
  return res.data;
};

export const createWorker = async (payload: WorkerCreatePayload): Promise<WorkerItem> => {
  const res = await apiClient.post<WorkerItem>('/workers', payload);
  return res.data;
};

export const updateWorker = async (id: string, payload: WorkerUpdatePayload): Promise<WorkerItem> => {
  const res = await apiClient.patch<WorkerItem>(`/workers/${id}`, payload);
  return res.data;
};

// ---------------------------------------------------------------------------
// Statutory Labour Attendance Endpoints
// ---------------------------------------------------------------------------

export const fetchAttendance = async (params?: {
  mine_site_id?: string;
  is_violation?: boolean;
  query?: string;
  worker_id?: string;
  limit?: number;
}): Promise<AttendanceItem[]> => {
  const res = await apiClient.get<AttendanceItem[]>('/labour/attendance', { params });
  return res.data;
};

export const createAttendance = async (payload: AttendanceCreatePayload): Promise<AttendanceItem> => {
  const res = await apiClient.post<AttendanceItem>('/labour/attendance', payload);
  return res.data;
};

export const bulkCreateAttendance = async (payload: AttendanceBulkPayload): Promise<AttendanceItem[]> => {
  const res = await apiClient.post<AttendanceItem[]>('/labour/attendance/bulk', payload);
  return res.data;
};

export const fetchLabourViolations = async (mine_site_id?: string): Promise<LabourViolationsSummary> => {
  const res = await apiClient.get<LabourViolationsSummary>('/labour/violations', {
    params: mine_site_id ? { mine_site_id } : undefined,
  });
  return res.data;
};

export const fetchLabourRules = async (mine_site_id?: string): Promise<LabourRule> => {
  const res = await apiClient.get<LabourRule>('/labour/rules', {
    params: mine_site_id ? { mine_site_id } : undefined,
  });
  return res.data;
};
